trie/tracer_test.go

```diff
@@ -0,0 +1,370 @@
+		t.Fatalf("Invalid accessList %v", err)
+	}
+}
+
+// Tests origin values won't be tracked in Iterator or Prover
+func TestAccessListLeak(t *testing.T) {
+	var (
+		db   = NewDatabase(rawdb.NewMemoryDatabase())
+		trie = NewEmpty(db)
+	)
+	// Create trie from scratch
+	for _, val := range standard {
+		trie.Update([]byte(val.k), []byte(val.v))
+	}
+	root, nodes, _ := trie.Commit(false)
+	db.Update(NewWithNodeSet(nodes))
+
+	var cases = []struct {
+		op func(tr *Trie)
+	}{
+		{
+			func(tr *Trie) {
+				it := tr.NodeIterator(nil)
+				for it.Next(true) {
+				}
+			},
+		},
+		{
+			func(tr *Trie) {
+				it := NewIterator(tr.NodeIterator(nil))
+				for it.Next() {
+				}
+			},
+		},
+		{
+			func(tr *Trie) {
+				for _, val := range standard {
+					tr.Prove([]byte(val.k), 0, rawdb.NewMemoryDatabase())
+				}
+			},
+		},
+	}
+	for _, c := range cases {
+		trie, _ = New(TrieID(root), db)
+		n1 := len(trie.tracer.accessList)
+		c.op(trie)
+		n2 := len(trie.tracer.accessList)
+
+		if n1 != n2 {
+			t.Fatalf("AccessList is leaked, prev %d after %d", n1, n2)
+		}
+	}
+}
+
+// Tests whether the original tree node is correctly deleted after being embedded
+// in its parent due to the smaller size of the original tree node.
+func TestTinyTree(t *testing.T) {
+	var (
+		db   = NewDatabase(rawdb.NewMemoryDatabase())
+		trie = NewEmpty(db)
+	)
+	for _, val := range tiny {
+		trie.Update([]byte(val.k), randBytes(32))
+	}
+	root, set, _ := trie.Commit(false)
+	db.Update(NewWithNodeSet(set))
+
+	trie, _ = New(TrieID(root), db)
+	orig := trie.Copy()
+	for _, val := range tiny {
+		trie.Update([]byte(val.k), []byte(val.v))
+	}
+	root, set, _ = trie.Commit(false)
+	db.Update(NewWithNodeSet(set))
+
+	trie, _ = New(TrieID(root), db)
+	if err := verifyAccessList(orig, trie, set); err != nil {
+		t.Fatalf("Invalid accessList %v", err)
+	}
+}
+
+func compareSet(setA, setB map[string]struct{}) bool {
+	if len(setA) != len(setB) {
+		return false
+	}
+	for key := range setA {
+		if _, ok := setB[key]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func forNodes(tr *Trie) map[string][]byte {
+	var (
+		it    = tr.NodeIterator(nil)
+		nodes = make(map[string][]byte)
+	)
+	for it.Next(true) {
+		if it.Leaf() {
+			continue
+		}
+		blob, _ := tr.reader.nodeBlob(it.Path(), it.Hash())
+		nodes[string(it.Path())] = common.CopyBytes(blob)
+	}
+	return nodes
+}
+
+func iterNodes(db *Database, root common.Hash) map[string][]byte {
+	tr, _ := New(TrieID(root), db)
+	return forNodes(tr)
+}
+
+func forHashedNodes(tr *Trie) map[string][]byte {
+	var (
+		it    = tr.NodeIterator(nil)
+		nodes = make(map[string][]byte)
+	)
+	for it.Next(true) {
+		if it.Hash() == (common.Hash{}) {
+			continue
+		}
+		blob, _ := tr.reader.nodeBlob(it.Path(), it.Hash())
+		nodes[string(it.Path())] = common.CopyBytes(blob)
+	}
+	return nodes
+}
+
+func diffTries(trieA, trieB *Trie) (map[string][]byte, map[string][]byte, map[string][]byte) {
```

Comment:
Could u add description here, like 
`Return the diff and shared nodes between 2 tries`

Reply:
Added